presto-main/src/main/java/com/facebook/presto/sql/planner/Partitioning.java

```diff
@@ -242,42 +250,78 @@ public String toString()
+    public static final class Translator
+    {
+        private final Function<Symbol, Optional<Symbol>> columnTranslator;
+        private final Function<Symbol, Optional<NullableValue>> constantTranslator;
+        private final Function<Expression, Optional<Symbol>> expressionTranslator;
+
+        public Translator(
+                Function<Symbol, Optional<Symbol>> columnTranslator,
+                Function<Symbol, Optional<NullableValue>> constantTranslator,
+                Function<Expression, Optional<Symbol>> expressionTranslator)
+        {
+            this.columnTranslator = requireNonNull(columnTranslator, "columnTranslator is null");
+            this.constantTranslator = requireNonNull(constantTranslator, "constantTranslator is null");
+            this.expressionTranslator = requireNonNull(expressionTranslator, "expressionTranslator is null");
+        }
+    }
+
     @Immutable
     public static final class ArgumentBinding
     {
         private final Symbol column;
         private final NullableValue constant;
+        private final Expression expression;
 
         @JsonCreator
         public ArgumentBinding(
                 @JsonProperty("column") Symbol column,
-                @JsonProperty("constant") NullableValue constant)
+                @JsonProperty("constant") NullableValue constant,
+                @JsonProperty("coalesce") Expression expression)
         {
             this.column = column;
             this.constant = constant;
-            checkArgument((column == null) != (constant == null), "Either column or constant must be set");
+            this.expression = expression;
+            checkArgument((column != null) || (constant != null) || (expression != null), "Either column, constant or expression must be set");
         }
 
         public static ArgumentBinding columnBinding(Symbol column)
         {
-            return new ArgumentBinding(requireNonNull(column, "column is null"), null);
+            return new ArgumentBinding(requireNonNull(column, "column is null"), null, null);
         }
 
         public static ArgumentBinding constantBinding(NullableValue constant)
         {
-            return new ArgumentBinding(null, requireNonNull(constant, "constant is null"));
+            return new ArgumentBinding(null, requireNonNull(constant, "constant is null"), null);
+        }
+
+        public static ArgumentBinding expressionBinding(Expression expression)
+        {
+            return new ArgumentBinding(null, null, requireNonNull(expression, "expression is null"));
         }
 
         public boolean isConstant()
         {
             return constant != null;
         }
 
+        public boolean isExpression()
+        {
+            return expression != null;
+        }
+
         public boolean isVariable()
         {
             return column != null;
         }
 
+        public boolean isColumnReference()
+        {
+            return column != null || expression != null;
```

Comment:
If expression is anything other than SymbolReference, it’s not a column reference

Reply:
Also, isVariable and isColumnReference should be the same method. It's fine to keep either name.

Reply:
If by `isColumnReference` you mean `isVariable` then the logic you mentioned would not work. It should be `noNonColumnReferenceExpression && partitionedOnColumns`.